### PR TITLE
Add missing parameter to `TableBorder.symmetric`, and improve class constructors

### LINKED ISSUE #144277: `TableBorder.symmetric` is missing the `borderRadius` parameter
This should be a very easy fix.

I also realized that even though `TableBorder.symmetric` is currently set up as a factory, it could instead be a `const` constructor.

### PR DESCRIPTION
**Pull request by nate-thegrate**
Originally, my aim was just to refactor (as per usual), but while messing around with the `TableBorder.symmetric` constructor, I realized that `borderRadius` was missing!

This pull request makes a few class constructors more efficient, and it fixes #144277 by adding the missing parameter.

<br>

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

File: dev/devicelab/lib/framework/browser.dart [hunk 1]
```diff
@@ -422,18 +422,6 @@ Duration _computeAverageDuration(List<BlinkTraceEvent> events) {
...
-    required this.args,
-    required this.cat,
-    required this.name,
-    required this.ph,
-    this.pid,
-    this.tid,
-    this.ts,
-    this.tts,
-    this.tdur,
-  });
```
Comment (nate-thegrate) [inline dev/devicelab/lib/framework/browser.dart:435]: This constructor was only used in `BlinkTraceEvent.fromJson`, so it's no longer needed.

File: packages/flutter/lib/src/material/material_state.dart [hunk 2]
```diff
@@ -321,12 +321,6 @@ abstract class MaterialStateBorderSide extends BorderSide implements MaterialSta
   /// const constructors so that they can be used in const expressions.
   const MaterialStateBorderSide();
 
-  /// Returns a [BorderSide] that's to be used when a Material component is
-  /// in the specified state. Return null to defer to the default value of the
-  /// widget or theme.
-  @override
-  BorderSide? resolve(Set<MaterialState> states);
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/material_state.dart:328]: Some methods needed to be moved down due to the "constructors should come before other stuff" rule.

File: packages/flutter/lib/src/services/autofill.dart [hunk 3]
```diff
@@ -13,11 +13,7 @@ export 'text_input.dart' show TextEditingValue, TextInputClient, TextInputConfig
 /// Each hint is pre-defined on at least one supported platform. See their
 /// documentation for their availability on each platform, and the platform
 /// values each autofill hint corresponds to.
-class AutofillHints {
-  // This class is not meant to be instantiated or extended; this constructor
-  // prevents instantiation and extension.
-  AutofillHints._();
-
+abstract final class AutofillHints {
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/services/autofill.dart:16]: Most of these private constructors were removed once the `final` class modifier was added; I finished the rest in this PR.

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/media_query.dart [hunk 4]
```diff
@@ -190,7 +190,7 @@ class MediaQueryData {
     'This constructor was deprecated in preparation for the upcoming multi-window support. '
     'This feature was deprecated after v3.7.0-32.0.pre.'
   )
-  factory MediaQueryData.fromWindow(ui.FlutterView window) => MediaQueryData.fromView(window);
```
[the remark this answers is not in the document]
Comment (nate-thegrate) [reply packages/flutter/lib/src/widgets/media_query.dart:193]: I went down a "class constructor rabbit hole" today—apparently the act of assigning a factory constructor is barely documented at all!

I proposed a linter rule over at https://github.com/dart-lang/linter/issues/4909, and I also made #144501, so hopefully information on factory constructors will be more easily accessible moving forward.

File: packages/flutter_tools/lib/src/base/utils.dart [hunk 5]
```diff
@@ -152,9 +152,7 @@ class SettingsFile {
     }
   }
 
-  factory SettingsFile.parseFromFile(File file) {
-    return SettingsFile.parse(file.readAsStringSync());
-  }
+  SettingsFile.parseFromFile(File file) : this.parse(file.readAsStringSync());
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/base/utils.dart:155]: while it seems like this still does the right thing, I would say this code is very confusing. We are invoking a named constructor strictly for the side effects and discarding the return value?

Can you either revert this diff, or inline the body of `SettingsFile.parse` to this constructor body?

File: packages/flutter_tools/lib/src/platform_plugins.dart [hunk 6]
```diff
@@ -244,18 +239,14 @@ class IOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPlug
...
-      defaultPackage: yaml[kDefaultPackage] as String?,
-      sharedDarwinSource: yaml[kSharedDarwinSource] as bool?,
-    );
-  }
+  IOSPlugin.fromYaml(this.name, YamlMap yaml)
+      : assert(validate(yaml)), // TODO(zanderso): https://github.com/flutter/flutter/issues/67241
+        classPrefix = '',
+        pluginClass = yaml[kPluginClass] as String?,
+        dartPluginClass = yaml[kDartPluginClass] as String?,
+        ffiPlugin = yaml[kFfiPlugin] as bool? ?? false,
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/platform_plugins.dart:247]: why did you add a fallback to false here?

File: packages/flutter_tools/lib/src/platform_plugins.dart [hunk 7]
```diff
@@ -244,18 +239,14 @@ class IOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPlug
...
+        defaultPackage = yaml[kDefaultPackage] as String?,
+        sharedDarwinSource = yaml[kSharedDarwinSource] as bool? ?? false;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/platform_plugins.dart:249]: same question here?

File: packages/flutter_tools/lib/src/platform_plugins.dart [hunk 8]
```diff
@@ -321,22 +312,14 @@ class MacOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPl
...
-    );
-  }
+  MacOSPlugin.fromYaml(this.name, YamlMap yaml)
+      : assert(validate(yaml)),
+        // Treat 'none' as not present. See https://github.com/flutter/flutter/issues/57497.
+        pluginClass = yaml[kPluginClass] == 'none' ? null : yaml[kPluginClass] as String?,
+        dartPluginClass = yaml[kDartPluginClass] as String?,
+        ffiPlugin = yaml[kFfiPlugin] as bool? ?? false,
+        defaultPackage = yaml[kDefaultPackage] as String?,
+        sharedDarwinSource = yaml[kSharedDarwinSource] as bool? ?? false;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/platform_plugins.dart:322]: why the fallback?

**Review by christopherfujino — CHANGES_REQUESTED** [review]
some questions

File: packages/flutter_tools/lib/src/base/utils.dart (hunk 5 above)
Comment (goderbauer) [reply packages/flutter_tools/lib/src/base/utils.dart:155]: Just some context: This is a redirecting constructor, a (not very well known and hence underutilized) feature in dart:
* https://dart.dev/language/constructors#redirecting-constructors
* https://stackoverflow.com/questions/20015249/in-dart-can-you-call-another-constructor-from-a-constructor

The return value of the `this.parse` constructor is not discarded, it is the return value of the `SettingsFile.parseFromFile` constructor.

Of course, it's up to us to decide whether we want to use this feature in our code base or if we think it makes things harder to read. Personally, I am OK with it since it is not that much different from calling a super constructor:

```dart
Foo(int b) : super.namedConstructor(a: b);
```

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/base/utils.dart:155]: ahh, ok, I read about this at some point and promptly forgot. This is fine.

File: packages/flutter_tools/lib/src/platform_plugins.dart (hunk 6 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/platform_plugins.dart:247]: ok, makes sense.

**Review by christopherfujino — APPROVED** [review]
LGTM
